### ISSUE TITLE
Memory leak when training the network

### ISSUE THREAD

**Issue by Westerby** [AlvinYH/Faster-VoxelPose#2]
Dear Authors,

I tried to train the network, but the memory usage on GPU keeps increasing with every iteration.

My specs:
torch: 1.4.0
torchvision: 0.5.0
cuda: 10.1

Steps to reproduce:
1. Pull the repo
2. Changed the data path in the config to my local dataset
3. Download pretrained backbone from voxelpose repository
4. set GPUS: '0' in config
5. Data preprocessing runs (resizing images)
6. Run the training: python run/train.py --cfg configs/panoptic/jln64.yaml

I tried to pinpoint the exact spot where the memory leak occurs, but it seems like it happens in different places in the network. The major increase seem to happen:
- i the resnet backbone, when 1st pass out of 5 goes through self.layer1 
- after that a small one in encoder-decoder part of the pose-net.

I use GPUtil for checking, the memory, and put it in the following places:

`        if views is not None:
            input_heatmaps = torch.stack([self.backbone(view) for view in views], dim=0)
        else:
            input_heatmaps = torch.stack(input_heatmaps, dim=0)
        print("after self.backbone", GPUtil.getGPUs()[0].memoryUsed)
        batch_size = input_heatmaps[0].shape[0]
 
        # human detection network
        proposal_heatmaps_2d, proposal_heatmaps_1d, proposal_centers, \
                              bbox_preds = self.pose_net(input_heatmaps, meta)
        print("after self.pose_net", GPUtil.getGPUs()[0].memoryUsed)
        mask = (proposal_centers[:, :, 3] >= 0)
        
        # joint localization network
        fused_poses, poses = self.joint_net(meta, input_heatmaps, proposal_centers.detach(), mask)
        print("after self.joint_net", GPUtil.getGPUs()[0].memoryUsed)
        # compute the training loss
        if self.training:
            assert targets is not None, 'proposal ground truth not set'
            proposal2gt = proposal_centers[:, :, 3]
            proposal2gt = torch.where(proposal2gt >= 0, proposal2gt, torch.zeros_like(proposal2gt))

            # compute 2d loss of proposal heatmaps
            loss_2d = F.mse_loss(proposal_heatmaps_2d[:, 0], targets['2d_heatmaps'], reduction='mean')
            
            # unravel the 1d gt heatmaps and compute 1d loss
            matched_heatmaps_1d = torch.gather(targets['1d_heatmaps'], dim=1, index=proposal2gt.long()\
                                               .unsqueeze(2).repeat(1, 1, proposal_heatmaps_1d.shape[2]))
            loss_1d = F.mse_loss(proposal_heatmaps_1d[mask], matched_heatmaps_1d[mask], reduction='mean')
            
            # compute the loss of bbox regression, only apply supervision on gt positions
            bbox_preds = torch.gather(bbox_preds, 1, targets['index'].long().view(batch_size, -1, 1).repeat(1, 1, 2))
            loss_bbox = F.l1_loss(bbox_preds[targets['mask']], targets['bbox'][targets['mask']], reduction='mean')

            del proposal_heatmaps_2d, proposal_heatmaps_1d, bbox_preds
            
            # weighted L1 loss of joint localization
            joints_3d = torch.gather(meta[0]['joints_3d'].float(), dim=1, index=proposal2gt.long().view\
                                             (batch_size, -1, 1, 1).repeat(1, 1, self.num_joints, 3))[mask]
            joints_vis = torch.gather(meta[0]['joints_3d_vis'].float(), dim=1, index=proposal2gt.long().view\
                                             (batch_size, -1, 1).repeat(1, 1, self.num_joints))[mask].unsqueeze(2)
            loss_joint = F.l1_loss(poses[0][mask] * joints_vis, joints_3d[:, :, :2] * joints_vis, reduction="mean") +\
                         F.l1_loss(poses[1][mask] * joints_vis, joints_3d[:, :, ::2] * joints_vis, reduction="mean") +\
                         F.l1_loss(poses[2][mask] * joints_vis, joints_3d[:, :, 1:] * joints_vis, reduction="mean") +\
                         2 * F.l1_loss(fused_poses[mask] * joints_vis, joints_3d * joints_vis, reduction="mean")

            loss_dict = {
                "2d_heatmaps": loss_2d,
                "1d_heatmaps": loss_1d,
                "bbox": 0.1 * loss_bbox,
                "joint": loss_joint,
                "total": loss_2d + loss_1d + 0.1 * loss_bbox + loss_joint
            }
        else:
            loss_dict = None
        print("after lossess block", GPUtil.getGPUs()[0].memoryUsed)
        # confidence score
        fused_poses = torch.cat([fused_poses, proposal_centers[:, :, 3:5].reshape(batch_size,\
                                 -1, 1, 2).repeat(1, 1, self.num_joints, 1)], dim=3)
        return fused_poses, poses, proposal_centers.detach(), loss_dict, input_heatmaps`

Readings are as follows 1st iteration:

after self.backbone 2274.0   
after self.pose_net 1680.0   
after self.joint_net 1680.0   
after lossess block 1680.0

2nd iteration

after self.backbone 2630.0  
after self.pose_net 2640.0  
after self.joint_net 2640.0  
after lossess block 2640.0

3rd iteration

after self.backbone 2880.0  
after self.pose_net 2890.0  
after self.joint_net 2890.0  
after lossess block 2890.0

4th iteration

after self.backbone 3132.0
after self.pose_net 3140.0
after self.joint_net 3140.0
after lossess block 3142.0

And it goes on like this until OOM. 

Did you experience similar problems?
Thank you for your help.

**Comment by AlvinYH** [comment]
Thanks for your interest in our project. We have fixed the bug in the previous commits. Please check out the newest version and let me know if there is any problem.

**Comment by Westerby** [comment]
Thank you!
Looks like it works now.

**Comment by LiBinbinBUPT** [comment]
> Thank you! Looks like it works now.
> 
> By the way - can you tell what was the problem? I spent some time debugging this memory leak, and couldn't find the cause.

Hello, I want to ask some problem. Would you face with the problem that when the model starts training,the computer operates slowly and quit after it warning that the bounding box is not sufficiently large for few times?
